src/goto-programs/remove_instanceof.cpp

```diff
@@ -0,0 +1,255 @@
+    else
+      expr=big_or;
+  }
+  else
+  {
+    Forall_operands(it, expr)
+      lower_instanceof(*it, goto_program, this_inst, inst_switch);
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_instanceoft::lower_instanceof
+
+  Inputs: GOTO program instruction `target` whose instanceof expressions,
+          if any, should be replaced with explicit tests, and the
+          `goto_program` it is part of.
+
+ Outputs: Side-effect on `target` as above.
+
+ Purpose: See function above
+
+\*******************************************************************/
+
+void remove_instanceoft::lower_instanceof(
+  goto_programt &goto_program,
+  goto_programt::targett target,
+  instanceof_instt &inst_switch)
+{
+  auto old_size=inst_switch.size();
+  lower_instanceof(target->code, goto_program, target, inst_switch);
+  auto after_code_size=inst_switch.size();
+  lower_instanceof(target->guard, goto_program, target, inst_switch);
+  auto after_guard_size=inst_switch.size();
+  // The instruction-switching step below will malfunction if a check
+  // has been added for the code *and* for the guard. This should
+  // be impossible, because guarded instructions currently have simple
+  // code (e.g. a guarded-goto), but this assertion checks that this
+  // assumption is correct and remains true on future evolution of the
+  // allowable goto instruction types.
+  assert(old_size==after_code_size || after_code_size==after_guard_size);
+}
+
+/*******************************************************************\
+
+Function: remove_instanceoft::lower_instanceof
+
+  Inputs: `goto_program`, all of whose instanceof expressions will
+          be replaced by explicit class-identifier tests.
+
+ Outputs: Side-effect on `goto_program` as above.
+
+ Purpose: See function above
+
+\*******************************************************************/
+bool remove_instanceoft::lower_instanceof(goto_programt &goto_program)
+{
+  instanceof_instt inst_switch;
+  Forall_goto_program_instructions(target, goto_program)
+    lower_instanceof(goto_program, target, inst_switch);
+  if(!inst_switch.empty())
+  {
+    for(auto &p : inst_switch)
+    {
+      const goto_programt::targett &this_inst=p.first;
+      const goto_programt::targett &newinst=p.second;
+      this_inst->swap(*newinst);
+    }
+    goto_program.update();
+    return true;
+  }
+  else
+    return false;
+}
+
+/*******************************************************************\
+
+Function: remove_instanceoft::lower_instanceof
+
+  Inputs: None
+
+ Outputs: Side-effects on this->goto_functions, replacing every
+          instanceof in every function with an explicit test.
+
+ Purpose: See function above
+
+\*******************************************************************/
+
+void remove_instanceoft::lower_instanceof()
+{
+  bool changed=false;
+  for(auto &f : goto_functions.function_map)
+    changed|=lower_instanceof(f.second.body);
```

Comment:
In the past, @kroening has suggested that we shouldn't be using bit-wise operators on bool, and instead use `changed=lower_instanceof(f.second.body) || changed;`